File: flutter-app/lib/features/events/domain/event_status.dart
/// Event lifecycle states (EOS-002-P3-Part-04 §7). Wire values match the event-service.
enum EventStatus {
  draft,
  planning,
  venueReserved,
  vendorConfirmed,
  invitationsSent,
  rsvpCollection,
  payments,
  ready,
  liveEvent,
  completed,
  archived,
  cancelled;

  /// Backend wire value (snake_case).
  String get wire => switch (this) {
    EventStatus.draft => 'draft',
    EventStatus.planning => 'planning',
    EventStatus.venueReserved => 'venue_reserved',
    EventStatus.vendorConfirmed => 'vendor_confirmed',
    EventStatus.invitationsSent => 'invitations_sent',
    EventStatus.rsvpCollection => 'rsvp_collection',
    EventStatus.payments => 'payments',
    EventStatus.ready => 'ready',
    EventStatus.liveEvent => 'live_event',
    EventStatus.completed => 'completed',
    EventStatus.archived => 'archived',
    EventStatus.cancelled => 'cancelled',
  };

  String get label => switch (this) {
    EventStatus.draft => 'Draft',
    EventStatus.planning => 'Planning',
    EventStatus.venueReserved => 'Venue reserved',
    EventStatus.vendorConfirmed => 'Vendor confirmed',
    EventStatus.invitationsSent => 'Invitations sent',
    EventStatus.rsvpCollection => 'Collecting RSVPs',
    EventStatus.payments => 'Payments',
    EventStatus.ready => 'Ready',
    EventStatus.liveEvent => 'Live',
    EventStatus.completed => 'Completed',
    EventStatus.archived => 'Archived',
    EventStatus.cancelled => 'Cancelled',
  };

  static EventStatus fromWire(String value) => switch (value) {
    'draft' => EventStatus.draft,
    'planning' => EventStatus.planning,
    'venue_reserved' => EventStatus.venueReserved,
    'vendor_confirmed' => EventStatus.vendorConfirmed,
    'invitations_sent' => EventStatus.invitationsSent,
    'rsvp_collection' => EventStatus.rsvpCollection,
    'payments' => EventStatus.payments,
    'ready' => EventStatus.ready,
    'live_event' => EventStatus.liveEvent,
    'completed' => EventStatus.completed,
    'archived' => EventStatus.archived,
    'cancelled' => EventStatus.cancelled,
    _ => EventStatus.draft,
  };
}

File: backend-api/services/event-service/src/domain/event-status.ts
/**
 * Event lifecycle state machine (EOS-002-P3-Part-04 §7).
 *
 * The event's business lifecycle is stored in the document's `status` field. Progression
 * is forward-only through the pipeline, with `cancelled` reachable from any active state
 * and `archived` as the terminal, read-only state.
 */
export const EventStatus = {
  Draft: 'draft',
  Planning: 'planning',
  VenueReserved: 'venue_reserved',
  VendorConfirmed: 'vendor_confirmed',
  InvitationsSent: 'invitations_sent',
  RsvpCollection: 'rsvp_collection',
  Payments: 'payments',
  Ready: 'ready',
  LiveEvent: 'live_event',
  Completed: 'completed',
  Archived: 'archived',
  Cancelled: 'cancelled',
} as const;

export type EventStatus = (typeof EventStatus)[keyof typeof EventStatus];

/** Allowed forward transitions. `cancelled` is added to every active state below. */
const FORWARD: Partial<Record<EventStatus, EventStatus[]>> = {
  [EventStatus.Draft]: [EventStatus.Planning],
  [EventStatus.Planning]: [EventStatus.VenueReserved],
  [EventStatus.VenueReserved]: [EventStatus.VendorConfirmed],
  [EventStatus.VendorConfirmed]: [EventStatus.InvitationsSent],
  [EventStatus.InvitationsSent]: [EventStatus.RsvpCollection],
  [EventStatus.RsvpCollection]: [EventStatus.Payments],
  [EventStatus.Payments]: [EventStatus.Ready],
  [EventStatus.Ready]: [EventStatus.LiveEvent],
  [EventStatus.LiveEvent]: [EventStatus.Completed],
  [EventStatus.Completed]: [EventStatus.Archived],
  [EventStatus.Cancelled]: [EventStatus.Archived],
};

/** States from which an event may still be cancelled (everything active). */
const CANCELLABLE: readonly EventStatus[] = [
  EventStatus.Draft,
  EventStatus.Planning,
  EventStatus.VenueReserved,
  EventStatus.VendorConfirmed,
  EventStatus.InvitationsSent,
  EventStatus.RsvpCollection,
  EventStatus.Payments,
  EventStatus.Ready,
  EventStatus.LiveEvent,
];

export function isEventStatus(value: unknown): value is EventStatus {
  return typeof value === 'string' && (Object.values(EventStatus) as string[]).includes(value);
}

/** The set of states reachable from `from` in one transition. */
export function allowedTransitions(from: EventStatus): EventStatus[] {
  const forward = FORWARD[from] ?? [];
  const canCancel = CANCELLABLE.includes(from) ? [EventStatus.Cancelled] : [];
  return [...forward, ...canCancel];
}

export function canTransition(from: EventStatus, to: EventStatus): boolean {
  return allowedTransitions(from).includes(to);
}

/** Archived events are read-only (EOS-002-P3-Part-04 §24). */
export function isReadOnly(status: EventStatus): boolean {
  return status === EventStatus.Archived;
}

/** Event names are editable only until completion (EOS-002-P3-Part-04 §24). */
export function isNameEditable(status: EventStatus): boolean {
  return (
    status !== EventStatus.Completed &&
    status !== EventStatus.Archived &&
    status !== EventStatus.Cancelled
  );
}
